plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java

```diff
@@ -1508,6 +1516,55 @@ public void testUpdateAfterEqualityDelete()
         }
     }
 
+    @Test
+    public void testInsertWithAutoCleanMetadataFile()
+            throws IOException
+    {
+        int metadataPreviousVersionCount = 5;
+        String tableName = "table_to_metadata_count" + randomNameSuffix();
+        assertUpdate("CREATE TABLE " + tableName + " (_bigint BIGINT, _varchar VARCHAR)");
+        BaseTable icebergTable = loadTable(tableName);
+        icebergTable.updateProperties()
+                .set(METADATA_DELETE_AFTER_COMMIT_ENABLED, "true")
+                .set(METADATA_PREVIOUS_VERSIONS_MAX, String.valueOf(metadataPreviousVersionCount))
+                .commit();
+
+        TrinoFileSystem trinoFileSystem = fileSystemFactory.create(SESSION);
+        FileIterator fileIteratorOld = trinoFileSystem.listFiles(Location.of(icebergTable.location()));
+        Map<String, Long> historyMetadataFiles = new HashMap<>();
+        while (fileIteratorOld.hasNext()) {
+            FileEntry entry = fileIteratorOld.next();
+            if (entry.location().path().endsWith(".metadata.json")) {
+                TableMetadata tableMetadata = TableMetadataParser.read(icebergTable.io(), new ForwardingInputFile(trinoFileSystem.newInputFile(entry.location())));
+                historyMetadataFiles.put(entry.location().path(), tableMetadata.lastUpdatedMillis());
+            }
+        }
+
+        for (int i = 0; i < 10; i++) {
+            assertUpdate("INSERT INTO " + tableName + " VALUES (1, 'a')", 1);
+            Set<String> currentMetadataFiles = new HashSet<>();
+            FileIterator fileIteratorNew = trinoFileSystem.listFiles(Location.of(icebergTable.location()));
+            while (fileIteratorNew.hasNext()) {
+                FileEntry entry = fileIteratorNew.next();
+                if (entry.location().path().endsWith(".metadata.json")) {
```

Comment:
No need to iterate data directory. Please consider iterating only metadata directory.